Revise support and worker state callbacks

JamesWrigley opened the pull request:
There's a few changes in here, I would recommend reviewing each commit individually. The big ones are:
- Support for worker added/exiting/exited callbacks.
- Revise support (requires callbacks to implement it properly).

Depends on https://github.com/timholy/Revise.jl/pull/871.

jpsamaroo reviewed (approved):
Great stuff! I have some questions about the semantics of these callbacks w.r.t error conditions, but other than that, this all seems solid.

jpsamaroo commented:
One additional thought: should we have a way to detect a worker exit reason? If a worker exits due to a segfault or OOM, this might be important to know so that I can do some recovery actions or reporting to the user. It could also be something we add later, as its own set of callbacks, but I figure it's worth thinking about before merging this.

JamesWrigley commented on src/cluster.jl:
```diff
@@ -1185,7 +1276,7 @@ function deregister_worker(pg, pid)
             manage(w.manager, w.id, w.config, :deregister)
             if PGRP.topology !== :all_to_all || isclusterlazy()
-                for rpid in workers()
+                for rpid in filter(!=(myid()), workers())
```
No that only removes `pid` (the ID of the dead worker) from the internal list. `deregister_worker()` is called on the live workers when a worker `pid` goes down, so when we recurse our own ID will still be in `workers()`. This didn't cause issues before because the function can handle being called recursively, but now it would cause the `exited` callbacks to be called twice.

JamesWrigley commented on src/cluster.jl:
```diff
@@ -461,12 +461,14 @@ function addprocs(manager::ClusterManager; kwargs...)
+    new_workers = @lock worker_lock addprocs_locked(manager::ClusterManager; kwargs...)
+    for worker in new_workers
+        for callback in values(worker_added_callbacks)
+            callback(worker)
```
No I think we should definitely throw the exception so that it's obvious that worker initialization somehow failed, otherwise code that runs later assuming initialization succeeded may cause even more errors. But I did change how we run them in 90f44f6 so that they execute concurrently to not slow down `addprocs()` too much and so that we can have warnings about slow callbacks.

JamesWrigley commented on src/cluster.jl:
```diff
@@ -855,13 +857,86 @@ const HDR_COOKIE_LEN=16
+    add_worker_added_callback(f::Base.Callable; key=nothing)
+
+Register a callback to be called on the master process whenever a worker is
+added. The callback will be called with the added worker ID,
```
Good point, fixed in 90f44f6.

JamesWrigley commented on src/cluster.jl:
```diff
@@ -855,13 +857,86 @@ const HDR_COOKIE_LEN=16
+"""
+    remove_worker_added_callback(key)
+
+Remove the callback for `key`.
```
Fixed in 90f44f6.

JamesWrigley commented on src/cluster.jl:
```diff
@@ -855,13 +857,86 @@ const HDR_COOKIE_LEN=16
+
+Register a callback to be called on the master process whenever a worker is
+added. The callback will be called with the added worker ID,
+e.g. `f(w::Int)`. Returns a unique key for the callback.
```
Fixed in 90f44f6.

JamesWrigley commented on src/cluster.jl:
```diff
@@ -855,13 +857,86 @@ const HDR_COOKIE_LEN=16
+
+Register a callback to be called on the master process immediately before a
+worker is removed with [`rmprocs()`](@ref). The callback will be called with the
+worker ID, e.g. `f(w::Int)`. Returns a unique key for the callback.
```
Fixed in 90f44f6.

JamesWrigley commented on src/cluster.jl:
```diff
@@ -855,13 +857,86 @@ const HDR_COOKIE_LEN=16
+Register a callback to be called on the master process when a worker has exited
+for any reason (i.e. not only because of [`rmprocs()`](@ref) but also the worker
+segfaulting etc). The callback will be called with the worker ID,
+e.g. `f(w::Int)`. Returns a unique key for the callback.
```
Fixed in 90f44f6.

JamesWrigley commented on src/cluster.jl:
```diff
@@ -855,13 +857,86 @@ const HDR_COOKIE_LEN=16
+"""
+    remove_worker_exiting_callback(key)
+
+Remove the callback for `key`.
```
Fixed in 90f44f6.

JamesWrigley commented on src/cluster.jl:
```diff
@@ -855,13 +857,86 @@ const HDR_COOKIE_LEN=16
+"""
+    remove_worker_exited_callback(key)
+
+Remove the callback for `key`.
```
Fixed in 90f44f6.

JamesWrigley commented on src/cluster.jl:
```diff
@@ -1055,24 +1134,36 @@ julia> workers()
+        end
+
+        if timedwait(() -> all(istaskdone.(callback_tasks)), callback_timeout) === :timed_out
+            @warn "Some callbacks timed out, continuing to remove workers anyway"
```
Fixed in 90f44f6.

JamesWrigley commented on src/cluster.jl:
```diff
@@ -1218,6 +1309,14 @@ function deregister_worker(pg, pid)
+    # Call callbacks on the master
+    if myid() == 1
+        for callback in values(worker_exited_callbacks)
+            callback(pid)
```
Yeah that's better, added in 90f44f6.

JamesWrigley commented:
> One additional thought: should we have a way to detect a worker exit reason? If a worker exits due to a segfault or OOM, this might be important to know so that I can do some recovery actions or reporting to the user. It could also be something we add later, as its own set of callbacks, but I figure it's worth thinking about before merging this.

That is an excellent point :thinking: I think it should go with the current set of callbacks, e.g. what if the signature for the worker-exiting callbacks was `f(w::Int, reason::ExitReasonEnum)`?

JBlaschke commented:
Btw, I think this is really cool stuff @JamesWrigley! 

@jpsamaroo and I where discussing this in the context advanced scheduling (i.e. your compute might be growing or shrinking) -- or you're on shared nodes where the sysadmin might kill a worker which is leaking memory. I this context I've been building "nanny workers" that will re-add dead workers, etc. But this has been a pain in the ^%#@

I am happy with this PR as is, but fi there are spare cycles, I want propose additional improvements:
1. Instead of the `{added, exiting, exited}` states, I propose using `{starting, started, exiting, exited}`. I.e follow the semantics of `event => conclusion`. So `request was made to add a worker (starting) => the worker is ready for work (started)`; and `request was made to shut down a worker (exiting) => worker is gone (exited)`. This brings the API in line with resource managers like Slurm (which have distinct "I've started doing some" and "I'm done doing that thing" phases).
2. Is there a way to more effectively capture the state of an `exiting` worker? I.e. capturing the exit code is a step in the right direction, but perhaps we also want to capture a little more context that will help an application automatically recover from an unexpected failure (e.g. "working on dataset 23", or `/path/to/coredump`). I.e. have the user be able to define additional optional input args to some of the callbacks?

JamesWrigley commented:
Thanks for taking a look :)

About those things:
1. I don't have any objection to adding a `started` callback, but what would it be used for? I can't think of a situation where a library that isn't calling `addprocs()` itself (in which case it would already know which workers are being started) would need to know about a worker it can't connect to yet.
2. In the case of the worker being killed gracefully I think this could be implemented already by an `exiting` callback to extract some status information from the worker. But if it died from e.g. a segfault/OOM it gets a bit more complicated :thinking: Checking for a coredump would involve adding another worker on the same node (which might not even be possible if its slurm allocation is expired), and that's a bit more side-effects than I'm comfortable with. Maybe we could pass the last ~100 lines of the logs to the callback?

   Alternatively, we could support some kind of per-worker status that a worker can set for itself (e.g. `setmystatus("working on dataset 23")`), and pass the last status to the `exited` callback?

JBlaschke commented:
> * I don't have any objection to adding a `started` callback, but what would it be used for? I can't think of a situation where a library that isn't calling `addprocs()` itself (in which case it would already know which workers are being started) would need to know about a worker it can't connect to yet.

In general I find it harder to change an API after the fact ;) -- Anyway the situation you're describing is where one worker is responsible for managing the overall workflow, as is common. However this is not always the case at scale (eg. what if we had >10k workers?). In those situations you often lay out your workers on a tree with several "management" nodes (eg. think fat trees, but for workers and not networks). In this situation you want to build callbacks that notify those manager nodes that the leaves have just changed (or are about to change).

> * In the case of the worker being killed gracefully I think this could be implemented already by an `exiting` callback to extract some status information from the worker. But if it died from e.g. a segfault/OOM it gets a bit more complicated 🤔 Checking for a coredump would involve adding another worker on the same node (which might not even be possible if its slurm allocation is expired), and that's a bit more side-effects than I'm comfortable with. Maybe we could pass the last ~100 lines of the logs to the callback?

Slurm can be configured to send a signal `N` seconds before killing a process. We also have the ability to add hooks to Slurm jobs upon process startup/completion. Oh, and coredumps don't have to land in `/tmp` -- I recommend `$SCRATCH` for these things at NERSC. Things like passing back part of the Slurm log is definitely in the right direction. I was thinking of something that the users and sysadmins can configure for a given library / application though.

>   Alternatively, we could support some kind of per-worker status that a worker can set for itself (e.g. `setmystatus("working on dataset 23")`), and pass the last status to the `exited` callback?

That's neat! I love it -- it can also help with overall workflow tooling.

All-in-all I am happy with this PR as is -- my comments are meant to make somethings that's great even better.

JamesWrigley commented:
Alrighty, switched to `{starting, started, exiting, exited}` callbacks in 8b04241 / 21da9fc.

> Slurm can be configured to send a signal `N` seconds before killing a process. We also have the ability to add hooks to Slurm jobs upon process startup/completion. Oh, and coredumps don't have to land in `/tmp` -- I recommend `$SCRATCH` for these things at NERSC. Things like passing back part of the Slurm log is definitely in the right direction. I was thinking of something that the users and sysadmins can configure for a given library / application though.

On second thoughts I'm unsure about promising logs because I don't know how the `RemoteLogger`'s thing will pan out: https://github.com/JuliaLang/Distributed.jl/pull/94. We can always promise to send the `stdout` of the worker because we're guaranteed to get that, but it might not have any useful information.

Am I correct in thinking that worker statuses and the worker-exited callbacks are sufficient for the uses you're talking about? The way I think about it is that there's three possible scenarios for exits:
1. Worker gets a request to exit (e.g. by slurm signal): it sets its status and then remotecalls the master to `rmproc()` itself. The worker-exited callback is called like `f(pid, last_status, ExitKind_graceful)`.
2. Worker is removed by the master: it may get a chance to set a final status if there's a worker-exiting callback that will tell it do so, either way the worker-exited callback gets called like `f(pid, last_status, ExitKind_graceful)`.
3. Worker dies suddenly (segfault/OOM etc): the worker-exited callback gets called like `f(pid, last_status, ExitKind_forceful)` and it can look for coredumps etc.

JamesWrigley commented:
Some updates:
- https://github.com/timholy/Revise.jl/pull/871 is merged and released now, but with a new minimum Julia version of 1.10 so I bumped the minimum Julia version for DistributedNext too in 914e50c.
- I took the liberty of rebasing since the commit history was getting a bit long.
- The worker-exited callbacks now get the worker state as well as their ID, which will tell them if the worker exited gracefully or not. As part of this I renamed the `WorkerState` instances and added a new state for forceful exits (`WorkerState_exterminated`) in d5fd837. The next thing to do is add support for worker statuses and pass those to the worker-exited callbacks, and then I think this will be ready to merge.

JamesWrigley commented:
Alrighty, implemented worker statuses in 64aba00. Now they'll be passed to the worker-exited callbacks.

Apologies for how big this PR is becoming :sweat_smile: I've tried to keep the commits atomic so you can review them one-by-one.

JamesWrigley commented:
Hmm, interestingly 0d5aaa3 seems to have almost entirely fixed https://github.com/JuliaParallel/DistributedNext.jl/issues/6. There are no more timeouts on Linux/OSX and I see only one on Windows.

The common problem with these hangs seems to be lingering tasks blocking Julia from exiting. At some point we should probably audit Distributed[Next] to remove all of them.

JamesWrigley commented:
(bump)

jpsamaroo commented on src/cluster.jl:
```diff
@@ -1041,14 +1209,67 @@ Identical to [`workers()`](@ref) except that the current worker is filtered out.
 """
 other_workers() = filter(!=(myid()), workers())
 
+"""
```
I'm a little concerned about the composability of this mechanism. If a library wants to set a status for debugging how it interacts with workers, that would conflict with the user or another library trying to do the same thing. Maybe we can also pass a key and track multiple statuses? I've used the calling `Module` in the past as a key (usually automated with a macro) when a package-global key is needed for a system like this.

I know my proposal isn't as convenient for worker-exiting callbacks to handle, so I'm open to alternatives, but I do think we'll need to consider the usability of this mechanism before merging it.

jpsamaroo reviewed (changes requested):
Awesome changes, I think the newest implementation is really clear! I do have concerns about `setstatus/getstatus`, but otherwise everything looks good.

JamesWrigley commented on src/cluster.jl:
```diff
@@ -852,16 +870,166 @@ const LPROC = LocalProcess()
+
+!!! warning
+    Adding workers can fail so it is not guaranteed that the workers requested
+    will exist.
```
It means that if e.g. 2 workers are requested on specific nodes they may not actually exist in the future since adding them may fail if the nodes are unreachable or something. I'll reword that to make it more clear.

JamesWrigley replied:
Fixed in a9f5dc8.

JamesWrigley commented on src/cluster.jl:
```diff
@@ -852,16 +870,166 @@ const LPROC = LocalProcess()
+    end
+
+    # Wait on the tasks so that exceptions bubble up
+    wait.(values(callback_tasks))
```
Fixed in a9f5dc8.

JamesWrigley commented on src/cluster.jl:
```diff
@@ -852,16 +870,166 @@ const LPROC = LocalProcess()
+"""
+    add_worker_started_callback(f::Base.Callable; key=nothing)
+
+Register a callback to be called on the master process whenever a worker is
```
Fixed in a9f5dc8.

JamesWrigley commented on src/cluster.jl:
```diff
@@ -852,16 +870,166 @@ const LPROC = LocalProcess()
+if `key` is not specified.
+
+All worker-exiting callbacks will be executed concurrently and if they don't
+all finish before the `callback_timeout` passed to `rmprocs()` then the process
```
Agreed, I went through all the docstrings and replaced `process` with `worker` in a9f5dc8.

JamesWrigley commented on src/cluster.jl:
````diff
@@ -1041,14 +1209,67 @@ Identical to [`workers()`](@ref) except that the current worker is filtered out.
+"working on dataset 42"
+```
+"""
+function setstatus(x, pid::Int=myid())
````
To me this is in the same class of function as the IO methods so it's not necessary, but I don't feel very strongly about it and it is more explicit this way. Renamed in 8ebd9bf.
